prov/rxm/src/rxm_ep.c

```diff
@@ -512,13 +516,21 @@ static ssize_t rxm_ep_send_common(struct fid_ep *ep_fid, const struct iovec *iov
 		pkt_size = sizeof(*pkt) + pkt->hdr.size;
 	}
 
-	ret = fi_send(rxm_conn->msg_ep, pkt, pkt_size, desc_tx_buf, 0, tx_entry);
-	if (ret) {
-		FI_WARN(&rxm_prov, FI_LOG_EP_DATA, "fi_send for MSG provider failed\n");
-		goto err;
+	if (flags & FI_INJECT & ~FI_COMPLETION) {
+		ret = fi_inject(rxm_conn->msg_ep, pkt, pkt_size, 0);
```

Comment:
Yes, FI_COMPLETION is handled in fi_cq_read family functions. But if app based on RXM prov has a deal with send that are expect no completion event (for instance, as fi_inject), buffer that are borrowed from pool won't be never freed, because user won't call fi_cq_read. In my case, I release borrowed buffer right after fi_inject.

I guess, fi_inject suspect more optimized implementation than fi_sendmsg(..., FI_INJECT). for example, fewer number of checks before buffer will be passed to transmission media.

Thank you for a good points.